app/src/main/java/com/setjy/practiceapp/presentation/ui/channels/middleware/SearchStreamsMiddleware.kt

```diff
@@ -0,0 +1,68 @@
+package com.setjy.practiceapp.presentation.ui.channels.middleware
+
+import com.setjy.practiceapp.presentation.base.mvi.Middleware
+import com.setjy.practiceapp.presentation.base.recycler.base.ViewTyped
+import com.setjy.practiceapp.presentation.ui.channels.ChannelsAction
+import com.setjy.practiceapp.presentation.ui.channels.ChannelsState
+import com.setjy.practiceapp.presentation.ui.channels.StreamItemUI
+import com.setjy.practiceapp.presentation.ui.channels.TopicItemUI
+import io.reactivex.rxjava3.core.Observable
+
+class SearchStreamsMiddleware : Middleware<ChannelsState, ChannelsAction> {
+    override fun bind(
+        actions: Observable<ChannelsAction>,
+        state: Observable<ChannelsState>
+    ): Observable<ChannelsAction> {
+        return actions.ofType(ChannelsAction.SearchStreams::class.java)
+            .distinct()
+            .withLatestFrom(state) { action, lastState ->
+                ChannelsAction.ShowSearchResult(
+                    if (action.query.isNotEmpty()) {
+                        val mutableItems = mutableListOf<ViewTyped>()
+                        putExpandedStreamsWithTopics(lastState.streams.orEmpty(), mutableItems)
+                        filterItems(mutableItems, action)
+                    } else {
+                        lastState.streams.orEmpty().map { it.copy(isExpanded = false) }
```

Comment:
lastState.streams.orEmpty().flatMap { stream ->
                            if (stream.isExpanded) {
                                listOf(stream) + stream.listOfTopics
                            } else {
                                listOf(stream)
                            }
                        }.filter {
                            when (it) {
                                is StreamItemUI -> {
                                    it.streamName.contains(action.query, true)
                                }
                                is TopicItemUI -> {
                                    it.topicName.contains(action.query, true)
                                }
                                else -> true
                            }
                        }